lazy_loader/__init__.py

```diff
@@ -257,15 +271,72 @@ def attach_stub(package_name: str, filename: str):
         incorrectly (e.g. if it contains an relative import from outside of the module)
     """
     stubfile = (
-        filename if filename.endswith("i") else f"{os.path.splitext(filename)[0]}.pyi"
+        filename if filename.endswith("i")
+        else f"{os.path.splitext(filename)[0]}.pyi"
     )
 
     if not os.path.exists(stubfile):
-        raise ValueError(f"Cannot load imports from non-existent stub {stubfile!r}")
+        raise ValueError(
+            f"Cannot load imports from non-existent stub {stubfile!r}",
+        )
 
     with open(stubfile) as f:
         stub_node = ast.parse(f.read())
 
     visitor = _StubVisitor()
     visitor.visit(stub_node)
     return attach(package_name, visitor._submodules, visitor._submod_attrs)
+
+
+class LazyFinder(MetaPathFinder):
+
+    def find_spec(
+        self,
+        fullname: str,
+        path: Sequence[str] | None,
+        target: types.ModuleType | None = ...,
+        / ,
+    ) -> ModuleSpec | None:
+        if not inside_context_manager.get():
+            # We are not in context manager, delegate to normal import
+            return None
+
+        if searching_spec.get():
+            # We are searching for the loader, so we should continue the search
+            return None
+
+        searching_spec.set(True)
+        spec = importlib.util.find_spec(fullname)
+        searching_spec.set(False)
+
+        if spec is None:
+            raise ModuleNotFoundError(f"No module named '{fullname}'")
+
+        spec.loader = importlib.util.LazyLoader(spec.loader)
+
+        return spec
+
+
+sys.meta_path.insert(0, LazyFinder())
```

Comment:
Most meta path finders will be targeted at specific set of modules. In those cases it is probably fine to insert the finder like this because in general if there are multiple finders/loaders affecting the same set of modules and injecting specialized behavior there is no way to guarantee that they can both do so safely anyway. In this case, however, we need the LazyFinder to respect other finders. If a finder is inserted into the meta path after this finder then it could preempt the lazy find. To address that, instead of inserting the finder here we should insert it inside the `__enter__` of the context manager and then remove it in `__exit__` with something like
```
with threading.Lock():
    sys.meta_path = [f for f in sys.meta_path if not isinstance(finder, LazyFinder)]
```
We will need to lock meta_path modification (both insertion and deletion) for thread safety. Since the `lazy_imports` context is by construction not reentrant it should be fine to not include any logic to handle the object already being in the list.

Reply:
Sorry for not answering earlier, as I was on vacation.

I know that it is possible for another hook to be inserted before this one after this is imported. However, that would only happen if it is explicitly inserted at the beginning of the list, which is not very common.

Although ensuring that we are at the first position on `__enter__` seems sensible at first glance, in a multithread context this simply cannot be done safely, AFAIK, because hooks are global and not per-thread, so you would be messing with the meta path for other threads too (and other threads could change it too!).

Your code seems broken. Not only it assumes that all threads would synchronize the meta path using the same lock (impracticable), but if I am not mistaken you are creating a new lock every time, so you are not even protecting anything (see https://github.com/pylint-dev/pylint/issues/5208).

Reply:
No problem!

Apologies, my example wasn't intended to be taken verbatim. Concretely, I would suggest storing a lock either at module scope or as a class member of the LazyFinder (e.g. upon first import of the `lazy_loader` itself). Then you would have a single instance that is shared across all threads and all import calls. I am confused as to why you think having all threads synchronizing the meta path via the same lock would be impractical. In general, thread-safe code _requires_ that all threads would use the same lock, otherwise the lock serves no purpose. Perhaps I am misunderstanding what you mean? Are you concerned that imports might be occurring in a performance-sensitive context? My assumption is that even if you have dozens/hundreds/more threads, the import shouldn't be on the critical path for performance.

Reply:
No, what I mean is that you cannot control what other threads do. In a multithreading context some threads may be created by other libraries or the application. These could in principle access the list while you are checking/changing it, and they **won't** use your lock. You can only protect by locking the APIs that you define, not the APIs defined elsewhere, unless you are in complete control of all threads of the process.

Reply:
Ah yes, if any other library includes modification of `sys.meta_path` beyond simply usage of this context manager in different threads then I agree that we have no control over that. There is nothing that we can do in that case. The question to me is really which case has more risk; allowing the lazy finder to be preempted by different insertion order, or allowing the lazy finder to have possible nondeterministic behavior in a multithreaded scenario where multiple distinct code paths modify `sys.meta_path`. I was assuming the latter would be _far_ less likely than the former (granting that the former is also quite unlikely) and was largely operating under the assumption that the lazy loader would be the only code modifying `sys.meta_path` in a multithreaded context, but the latter is certainly a worse problem in the sense that the nondeterministic code is a far greater evil and I wasn't really thinking much about it. I'm fine with leaving the implementation as is and just accepting the limitation.

Reply:
What I think we could do mostly safely is to warn if we are not in first position, possibly even listing which other hooks are before ours, so that users are aware of that and do not spend hours to debug it.

Reply:
Yeah, worst case with the warning is that we report a false positive if some other libraries is inserted and then removed with bad timing in another thread.